docs/source/migration/3.0.mdx

```diff
@@ -275,13 +282,47 @@ val request = ApolloRequest(query).withHttpFetchPolicy(CacheFirst)
 val response = apolloClient.query(request)
 ```
 
-## Optional
+### `CacheKeyResolver`
 
-Apollo Android distinguishes between null and absent values.
+The `CacheKeyResolver` API has been updated to work with declarative cache.
 
-Apollo Android 2.x uses `Input` to represent optional (maybe nullable) values.
+* `fromFieldRecordSet` is renamed to `cacheKeyForObject`.
+* `fromFieldArguments` is renamed to `cacheKeyForField`.
+* The `CacheKey` return value is now nullable, and `CacheKey.NONE` is replaced with `null`.
 
-Apollo Android 3.x uses `Optional` instead so that it can be used in other places than input types. For an example, fields could be made optional with an [`@optional`](https://github.com/apollographql/apollo-android/blob/5c1ff82ccaded93ae7126cedc6d7835e63fed75a/apollo-compiler/src/main/resources/apollo.graphqls#L6) directive. `Optional` is a sealed class so `when` statements don't need an `else` branch.
+```kotlin
+// Replace
+val resolver: CacheKeyResolver = object : CacheKeyResolver() {
+  override fun fromFieldRecordSet(field: ResponseField, recordSet: Map<String, Any>): CacheKey {
+    return CacheKey.from(recordSet["id"] as String)
+  }
+  override fun fromFieldArguments(field: ResponseField, variables: Operation.Variables): CacheKey {
+    return CacheKey.from(field.resolveArgument("id", variables) as String)
+  }
+}
+
+// With
+val cacheKeyResolver = object : CacheKeyResolver() {
+  override fun cacheKeyForObject(type: CompiledNamedType, obj: Map<String, Any?>): CacheKey? {
+    return (obj["id"] as? String)?.let { CacheKey(it) }
+  }
+  override fun cacheKeyForField(field: CompiledField, variables: Executable.Variables): CacheKey? {
+    return (field.resolveArgument("id", variables) as? String)?.let { CacheKey(it) }
+  }
+}
+```
+
+## Optional values
+
+### The `Optional` class
+
+Apollo Android distinguishes between `null` values and absent values.
+
+Apollo Android 2.x uses `Input` to represent optional (maybe nullable) values for input types.
+
+Apollo Android 3.x uses `Optional` instead so that it can be used in places besides input types. For example, fields can be made optional with an [`@optional`](https://github.com/apollographql/apollo-android/blob/5c1ff82ccaded93ae7126cedc6d7835e63fed75a/apollo-compiler/src/main/resources/apollo.graphqls#L6) directive.
```

Comment:
> fields can be made optional

I used conditional here because it's still very experimental. It was an illustration of things that could be done but are not really set in stone. Maybe it's better to just remove it ?

Reply:
ahh ok I will try simplifying this and we can re-elaborate if it feels better